spec.html

```diff
@@ -27557,122 +27812,85 @@ <h1>
-            _referencingScriptOrModule_: a Script Record, a Module Record, or *null*,
+          GetImportedModule (
+            _referrer_: a Cyclic Module Record,
             _specifier_: a String,
-          ): either a normal completion containing a Module Record or a throw completion
+          ): a Module Record
         </h1>
         <dl class="header">
           <dt>description</dt>
-          <dd>It provides the concrete Module Record subclass instance that corresponds to _specifier_ occurring within the context of the script or module represented by _referencingScriptOrModule_. _referencingScriptOrModule_ may be *null* if the resolution is being performed in the context of an <emu-xref href="#sec-import-calls">`import()`</emu-xref> expression and there is no active script or module at that time.</dd>
+          <dd></dd>
         </dl>
 
-        <emu-note>
-          <p>An example of when _referencingScriptOrModule_ can be *null* is in a web browser host. There, if a user clicks on a control given by</p>
-
-          <pre><code class="html">&lt;button type="button" onclick="import('./foo.mjs')"&gt;Click me&lt;/button&gt;</code></pre>
-
-          <p>there will be no active script or module at the time the <emu-xref href="#sec-import-calls">`import()`</emu-xref> expression runs. More generally, this can happen in any situation where the host pushes execution contexts with *null* ScriptOrModule components onto the execution context stack.</p>
-        </emu-note>
-
-        <p>An implementation of HostResolveImportedModule must conform to the following requirements:</p>
-        <ul>
-          <li>
-            If the returned Completion Record is a normal completion, it must be a normal completion containing an instance of a concrete subclass of Module Record.
-          </li>
-          <li>
-            If a Module Record corresponding to the pair _referencingScriptOrModule_, _specifier_ does not exist or cannot be created, an exception must be thrown.
-          </li>
-          <li>
-            Each time this operation is called, if it completes normally and the (_referencingScriptOrModule_, _specifier_, current Realm Record) triple is the same, it must return the same Module Record instance.
-          </li>
-        </ul>
-        <p>Multiple different _referencingScriptOrModule_, _specifier_ pairs may map to the same Module Record instance. The actual mapping semantic is host-defined but typically a normalization process is applied to _specifier_ as part of the mapping process. A typical normalization process would include actions such as alphabetic case folding and expansion of relative and abbreviated path specifiers.</p>
+        <emu-alg>
+          1. Let _record_ be the sole element of _referrer_.[[LoadedModules]] whose [[Specifier]] is _specifier_.
+          1. Assert: _record_ exists, because LoadRequestedModules has completed successfully on _referrer_ prior to invoking this abstract operation.
```

Comment:
The spec doesn't currently use this form:
```
1. Let _alias_ be [something that might or might not exist].
2. Assert: _alias_ exists. // If _alias_ exists ...
```
but maybe the editors are okay with it.

Reply:
I could use one of these:
- ```
  1. Let _record_ be ~empty~.
  2. If _referrer_.[[LoadedModules]] contains a Record whose [[Specifier]] is _specifier_, set _record_ to that Record.
  3. Assert: _record_ is not ~empty~.
  ```
  which reuses the "If ... contains X such that ... / Let _x_ be that X" wording from https://tc39.es/ecma262/#sec-privateelementfind
- ```
  1. Let _record_ be ~empty~.
  2. For each Record _r_ in _referrer_.[[LoadedModules]], do
     1. If _r_.[[Specifier]] is _specifier, then
        1. Assert: _record_ is ~empty~.
        2. Set _record_ to _r_.
  5. Assert: _record_ is not ~empty~.
  ```

Reply:
https://github.com/tc39/ecma262/pull/2877 changes PrivateElementFind to be a more explicit loop:

```
1. For each Private Name _pn_ of _names_, do
  1. If _pn_.[[Description]] is _identifier_, then
    1. Return _pn_.
```
But in that case you don't know it's _exactly_ one, because you might have to traverse the environment chain.

A more analogous case is [in the evaluation semantics for private names as class elements](https://tc39.es/ecma262/multipage/ecmascript-language-functions-and-classes.html#sec-class-definitions-runtime-semantics-evaluation):

```
4. Assert: Exactly one element of _names_ is a Private Name whose [[Description]] is _privateIdentifier_.
5. Let _privateName_ be the Private Name in _names_ whose [[Description]] is _privateIdentifier_.
```

That's what I'd do here:

```suggestion
          1. Assert: Exactly one element of _referrer_.[[LoadedModules]] is a Record whose [[Specifier]] is _specifier_, since LoadRequestedModules has completed successfully on _referrer_ prior to invoking this abstract operation.
          1. Let _record_ be the Record in _referrer_.[[LoadedModules]] whose [[Specifier]] is _specifier_.
```